aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -0,0 +1,309 @@
+    const IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const int64_t groups,
+    const float output_min,
+    const float output_max) {
+  const auto padding = expand_param_if_needed(padding_, "padding", 2);
+  const auto stride = expand_param_if_needed(stride_, "stride", 2);
+  const auto dilation = expand_param_if_needed(dilation_, "dilation", 2);
+  const Tensor weight_nhwc = weight.contiguous(MemoryFormat::ChannelsLast);
+
+  TORCH_CHECK(
+      available(
+          weight_nhwc,
+          bias,
+          padding,
+          stride,
+          dilation,
+          groups,
+          output_min,
+          output_max),
+      "xnnpack::convolution not available!");
+
+  xnn_operator_t convolution_op{};
+
+  const xnn_status create_status = xnn_create_convolution2d_nhwc_f32(
+      padding[Layout::Parameter::Height],                             // input_padding_top
+      padding[Layout::Parameter::Width],                              // input_padding_right
+      padding[Layout::Parameter::Height],                             // input_padding_bottom
+      padding[Layout::Parameter::Width],                              // input_padding_left
+      weight_nhwc.size(Layout::Filter::Height),                       // kernel_height
+      weight_nhwc.size(Layout::Filter::Width),                        // kernel_width
+      stride[Layout::Parameter::Height],                              // subsampling_height
+      stride[Layout::Parameter::Width],                               // subsampling_width
+      dilation[Layout::Parameter::Height],                            // dilation_height
+      dilation[Layout::Parameter::Width],                             // dilation_width
+      groups,                                                         // groups
+      weight_nhwc.size(Layout::Filter::Input),                        // group_input_channels
+      weight_nhwc.size(Layout::Filter::Output) / groups,              // group_output_channels
+      weight_nhwc.size(Layout::Filter::Input) * groups,               // input_pixel_stride
+      weight_nhwc.size(Layout::Filter::Output),                       // output_pixel_stride
+      weight_nhwc.data_ptr<float>(),                                  // kernel
+      (bias && bias->defined()) ? bias->data_ptr<float>() : nullptr,  // bias
+      output_min,                                                     // output_min
+      output_max,                                                     // output_max
+      0u,                                                             // flags
+      &convolution_op);                                               // operator
+
+  TORCH_CHECK(
+      xnn_status_success == create_status,
+      "xnn_create_convolution2d_nhwc_f32 failed!");
+
+  return Context{
+      Operator(convolution_op),
+      weight_nhwc.sizes().vec(),
+      padding,
+      stride,
+      dilation,
+  };
+}
+
+bool usable(const Tensor& input) {
```

Comment:
can we reuse the same shape checking from regular conv2d? my worry is that you'd produce very cryptic error message if the input has the wrong shape. Or does xnnpack return some kind of error string?

it's non-blocking, but please add a TODO for it